### ISSUE TITLE
Search match not immediately reported for newly delivered mail with fts_autoindex = no

### ISSUE THREAD

**Issue by edieterich** [slusarz/dovecot-fts-flatcurve#10]
I use `fts_autoindex = no`. When I deliver a small text mail and immediately search for it, the result is as expected, the mail is reported as a match. I'm using doveadm here for demonstration purposes, it's the same with IMAP.

```
doveadm save -u ewald testmail && doveadm -D search -u ewald mailbox inbox text test
Sep 30 13:30:15 doveadm(ewald): Debug: fts-flatcurve: Query (allhdrs:test* OR test*) mailbox=INBOX matches=1 uids=9
28795a3baf42f160202400006a82f8f2 9
```

Now I deliver a bigger mail. It has a plain text body and 2 images and 1 PDF as attachments. But I'm not indexing the attachments and the size of the plain text body is actually quite small.

```
doveadm save -u ewald testmail2 && doveadm -D search -u ewald mailbox inbox text anhang
Sep 30 13:38:08 doveadm(ewald): Debug: fts-flatcurve: Query (allhdrs:anhang* OR anhang*) mailbox=INBOX matches=1 uids=10
```

The debug log shows that flatcurve found a match, but it is not reported, there's no line with the mailbox-guid and the UID in the output. When I repeat the search, the match **is** reported:

```
doveadm -D search -u ewald mailbox inbox text anhang
Sep 30 13:40:11 doveadm(ewald): Debug: fts-flatcurve: Query (allhdrs:anhang* OR anhang*) mailbox=INBOX matches=1 uids=10
28795a3baf42f160202400006a82f8f2 10
```

Again, it's the same with IMAP. First search doesn't show any match, second one does:

```
. SEARCH TEXT anhang
* SEARCH
. OK Search completed (0.033 + 0.001 secs).
. SEARCH TEXT anhang
* SEARCH 1
. OK Search completed (0.001 + 0.000 secs).
```

The log shows a match for both searches:

```
Sep 30 13:43:03 imap(ewald)<34782><3nF5xDTNVMsAAAAAAAAAAAAAAAAAAAAB>: Debug: fts-flatcurve: Query (allhdrs:anhang* OR anhang*) mailbox=INBOX matches=1 uids=13
...
Sep 30 13:43:05 imap(ewald)<34782><3nF5xDTNVMsAAAAAAAAAAAAAAAAAAAAB>: Debug: fts-flatcurve: Query (allhdrs:anhang* OR anhang*) mailbox=INBOX matches=1 uids=13
```

Any idea what's going on here? Again, thanks for your help!

**Comment by slusarz** [comment]
With fts_autoindex = no and fts_enforced = no, there is an expectation that search results may not be immediately available.

But it is strange that it is listing a debug entry with (correct) matches.  I can't explain that.

You should try running the commands with debug logging on and post the output here.  "log_debug = category=fts" will probably provide better results than "mail_debug = yes" also.

**Comment by edieterich** [comment]
This appears to be a weird German language stemming problem, `fts_autoindex = no` is not the issue.

I wrote a test to reproduce the issue. I added a German word to the mbox file, set `fts_languages = de en` and added 2 scripts to search for an English and a German word. The German word in the mbox file is plural and I'm searching for the singular.

The sequence of commands is search, append, search, search. This works in English. The second search in German fails, the third one works.

Here's a diff of my changes: https://github.com/slusarz/dovecot-fts-flatcurve/compare/master...edieterich:reproduce-mail-not-found-after-append

Here's a test run: https://github.com/edieterich/dovecot-fts-flatcurve/runs/3767592751?check_suite_focus=true

This is the log line for failing second search after the append: https://github.com/edieterich/dovecot-fts-flatcurve/runs/3767592751?check_suite_focus=true#step:3:9788

`uids=1:2` looks good, but the test script only got the first mail as a match ("best match: SEARCH 1").

The log line for the third, succeeding search looks exactly the same: https://github.com/edieterich/dovecot-fts-flatcurve/runs/3767592751?check_suite_focus=true#step:3:9793

Can you make sense out of this? I don't know if this is a flatcurve problem or not.

**Comment by slusarz** [comment]
This is almost certainly related/duplicate of Issue #9.

The 1st German search:
```
Oct 03 17:59:29 imap(user)<1116><Ix7KjnbNsJh/AAAB>: Debug: fts-flatcurve: Last UID mailbox=imaptest uid=1
Oct 03 17:59:29 imap(user)<1116><Ix7KjnbNsJh/AAAB>: Debug: fts-flatcurve: Query (allhdrs:anhang* OR anhang*) mailbox=imaptest matches=1 uids=1:2
```
Because flatcurve announces to core fts code that it's highest known UID is 1, results > 1 will be ignored by the core fts code.  The 2nd time the search is run, the highest UID reported is 2, so all results are used.

FTS code *is* falling back to a manual search for UID 2 in the first example:
```
Oct 03 17:59:29 imap(user)<1116><Ix7KjnbNsJh/AAAB>: Debug: Mailbox imaptest: UID 2: Opened mail because: search
Oct 03 17:59:29 imap(user)<1116><Ix7KjnbNsJh/AAAB>: Debug: Mailbox imaptest: UID 2: Looked up field mime.parts from mail cache
```
... but I don't believe manual searching uses fts tokenizers, so that's why "anhang" is not matching "Anhänge".  (The same manual search happens in the English code ... but it is doing an exact match of "body" to the message contents, so fallback is working correctly there.)  I'll check with the team, but this is either a bug or limitation, albeit in core code not flatcurve.

Anyway, these tests correctly pass with the fixes from #9 : https://github.com/slusarz/dovecot-fts-flatcurve/actions/runs/1301035327 .  Can confirm that reverting to 3a3f16c7e246a7c946c633751e9a8b3994d02a6d causes the test to fail.

Since this is almost certainly duplicate of #9, will close this issue.
